### ISSUE TITLE
Set webagg.address, webagg.port and webagg.port_retries by environment variables

### ISSUE THREAD

**Issue by GreenCourt** [matplotlib/matplotlib#20550]
### Problem

- I use remote machines over ssh and use webagg to show plots.
    - Some of machines are shared with my colleagues.
- I install python packages in Docker containers.
- I want to change webagg.address, webadd.port and webagg.port_retries for each machine without hard-coding them in Docker images or python scripts.

### Proposed Solution

Read following environment variables to set the webagg parameters.

- MPLWEBAGG_IP
- MPLWEBAGG_PORT
- MPLWEBAGG_PORT_RETRIES

We can pass these environment variables to containers.

### Additional context and prior art

- jupyter can set the port by environment variables for containers.
    - https://github.com/jupyter/notebook/issues/5212
    - https://github.com/jupyter/notebook/pull/5221

- I will try to work on this if this feature is agreed.

**Comment by anntzer** [comment]
You can use a custom matplotlibrc file (although these admittedly won't compose with other matplotlibrc files); or, a more general solution would to be support setting $MATPLOTLIBRC to e.g. `webagg.ip=...,foo.bar=...` where the part after the "=" is a python literal (read by ast.literal_eval) (or a json if we prefer, etc.) and say that *if* $MATPLOTLIBRC contains an "=", then it is parsed as above and then merged with the matplotlibrc file.  (This avoids having to add new environment variables for each and every possible entry... and is composable as people can do `MATPLOTLIBRC=$MATPLOTLIBRC,additional.param=...,...`)

**Comment by timhoffm** [comment]
> and say that *if* $MATPLOTLIBRC contains an "=", then it is parsed as above

That sounds a bit too clever. Why not make one extra environment variable, say `MATPLOTLIBRC_DATA`.

**Comment by anntzer** [comment]
That would be fine for me too.

**Comment by GreenCourt** [comment]
Thanks for the response.

What format is preferred for `MATPLOTLIBRC_DATA`?

The format `webagg.address=...,foo.bar=...` maybe useful for users when adding a entry to existing `MATPLOTLIBRC_DATA` as @anntzer says. However a little complicated implementation and tests are required since the part after the "=" may contain "," or "=" as literal.

Json or python dictionary literal is easy to parse. But, for users, adding an entry is not easy.

I _personally_ do not need to add entries to existing `MATPLOTLIBRC_DATA`.
How do you think?

**Comment by timhoffm** [comment]
Quoting and separation cannot be parsed in bulk or using a regexp, but an iterative approach is reasonably simple:

1. Find next =
2. Everything till there is the key.
3. If the next char is a quote, find the next matching quote, else find the next comma.
4. Everything there is the next value.
5. If 3 was a quote, find the next comma.
6. Repeat from 1.


This is essentially: entries separated by comma with optionally quoted values.

**Comment by GreenCourt** [comment]
Can we assume that there is no escaped quotation or other edge cases?
Then we can implement like this (more test cases required).

```python3
import ast

MATPLOTLIBRC_DATA=' webagg.address=  "0.0.0.0" , webagg.port =12345 , webagg.port_retries = 12'

rc = {}

data = MATPLOTLIBRC_DATA.strip()
pos = data.find("=")
while pos != -1:
    key = data[:pos].strip()
    data = data[pos+1:].strip()

    if len(data) == 0:
        print("Warning message here.")
        break

    if data[0] == '"' or data[0] == "'":
        stop = data[1:].find(data[0]) # find the quotation pair
        if stop == -1:
            print("Warning message here.")
            break
        stop += 2
        assert(data[stop-1] == data[0]) # stop is at the charactor next to the quotation
    else:
        stop = data.find(",")
        if stop == -1:
            stop = len(data)
        assert(stop == len(data) or data[stop] == ",") # stop is at the comma

    rc[key] = ast.literal_eval(data[:stop].strip())
    data = data[stop:].strip()

    pos_comma = data.find(",")
    if pos_comma != -1:
        data = data[pos_comma+1:].strip()
         
    pos = data.find("=")

print(rc)
```

Another implementation is adding `{` and `}` to the `MATPLOTLIBRC_DATA` and parse it as a dictionary by ast.literal_eval.
In this case, users must quote the keys.

```python3
import ast
MATPLOTLIBRC_DATA='  "webagg.address"  : "0.0.0.0", "webagg.port":12345 ,'

rc = ast.literal_eval('{' + MATPLOTLIBRC_DATA + '}')
print(rc)
```

**Comment by anntzer** [comment]
The last approach seems nice.  If using the first approach, I guess the robust approach would be to parse using `ast.parse` and then walk the parse tree to replace the `foo.bar` assignment targets by `rcParams["foo.bar"]` (and thus use semicolons as separators).  But that's less nice to document, as opposed to "it's like what you put between the braces of a dict, except that you skip the braces".

**Comment by timhoffm** [comment]
The need for quoting is quite bad from a usability perspective.

**Comment by anntzer** [comment]
I don't agree: it would be a relatively advanced feature so it seems reasonable to sacrifice a *tiny* bit of syntax convenience in exchange for simplicity of implementation.  Plus, we're all used to quote python strings as well... (yes, I realize the situation is not *exactly* the same.)

**Comment by jklymak** [comment]
I'm pretty leery about this.  If we give folks _another_ way to edit the matplotlibrc they _will_ use it.  It would require pretty careful integration with the rest of the matpltolibrc machinery, which leads to confusion already, and we really should make sure the syntax is super-clear.

**Comment by timhoffm** [comment]
Is this really needed as a builtin feature? This should be quite rare and can be implemented by a handful lines of code for the originally requested rcParams: parse some environment variables and set the corresponding rcParams.

I propose to close rhis.

**Comment by tacaswell** [comment]
I agree to close this.  This level of config is best set by an application (which the user needs to write anyway to provide any plotting code) that is used to launch the service.
